### ISSUE TITLE
Try streaming dataframes with cudf

### ISSUE THREAD

**Issue by mrocklin** [python-streamz/streamz#222]
Ideally we would be able to reuse the `streamz.dataframe` module with pandas-like libraries, like `cudf`.

This would require removing explicit type checks like `isinstance(obj, pd.DataFrame)` with functional checks like `"DataFrame" in type(obj).__name__"` (or something better, like what is in `dask.dataframe.utils.is_dataframe_like`) and probably a bunch of other work.

However, if this works then ideally we would be able to reuse a bunch of code, and make `streamz.dataframe` a bit more generic.  

To start, I would probably make a random dataframe of pandas dataframes, and then convert them into cudf dataframes

```python
from streamz.dataframe import Random
import cudf

sdf = Random()
gdf = sdf.map_partitions(cudf.from_pandas, sdf)
gdf.x.sum()
```

Probably this will break for many reasons, both in streamz.dataframe and in cudf, but iterating on that set of problems might eventually yield something that is generic enough on both sides to work.

**Comment by mrocklin** [comment]
https://github.com/dask/dask/blob/9db0758bb3c9347f0dc42f726b31a63e6845d998/dask/dataframe/utils.py#L498-L511

**Comment by skmatti** [comment]
I have tried this with cudf and has been successful in making it work for some operations 
1. Creating and streaming cudf dataframes
2. Aggregations
3. Assigning new columns to the stream
4. Window operations with number of elements

But I am not able to get some of the below operations work, 
1. Window operations with time 
2. Rolling functions
3. GroupBy functions

cudf is missing a few implementations that does not permit above operations as of now

Some of the operations that needs to implemented in cudf to make the above mentioned operations to work:
1. DataFrame operations: sum, mean, add and count etc. These are implemented for Series but not for df yet. 
2. Rolling function
3. Datetime index seems to throw a casting [error](https://github.com/rapidsai/cudf/issues/835)
4. Concat along axis=1

**Comment by mrocklin** [comment]
Thanks for giving this a shot @skmatti and for reporting back.  I'm excited that you were able to get some things to work!  A few requests:

1.  I suspect that you had to change some code within the streamz library.  When you're ready, can you submit a pull request to this repository with those changes?  My guess is that we may have to iterate on them a bit
2.  For the cudf operations that are missing, could you raise issues at https://github.com/rapids/cudf ?  For example, here is an issue about reductions that you mention: https://github.com/rapidsai/cudf/issues/257

    In particular if you're able to isolate the extent of the Pandas API that you need that can help the cudf developers to build out that functionality.

**Comment by mrocklin** [comment]
Also cc @kkraus14, who I suspect will be interested in this

**Comment by martindurant** [comment]
Closed by #224
